contracts/src/core/libraries/v3/PaymentVaultLib.sol

```diff
@@ -0,0 +1,199 @@
+    }
+
+    struct QuorumConfig {
+        address token; // the address of the token used for on-demand payments.
+        address recipient; // the address of the recipient of the on-demand payments.
+        uint64 reservationSymbolsPerSecond;
+        uint64 onDemandSymbolsPerPeriod;
+        uint64 onDemandPricePerSymbol;
+    }
+
+    struct QuorumProtocolConfig {
+        uint64 minNumSymbols;
+        uint64 reservationAdvanceWindow;
+        uint64 reservationRateLimitWindow;
+        uint64 onDemandRateLimitWindow;
+        bool onDemandEnabled;
+    }
+
+    struct Reservation {
+        uint64 symbolsPerSecond;
+        uint64 startTimestamp;
+        uint64 endTimestamp;
+    }
+}
+
+library PaymentVaultStorage {
+    string internal constant STORAGE_ID = "eigen.da.payment.vault";
+    bytes32 internal constant STORAGE_POSITION =
+        keccak256(abi.encode(uint256(keccak256(abi.encodePacked(STORAGE_ID))) - 1)) & ~bytes32(uint256(0xff));
+
+    struct Layout {
+        mapping(uint64 => PaymentVaultTypes.Quorum) quorum;
+    }
+
+    function layout() internal pure returns (Layout storage s) {
+        bytes32 position = STORAGE_POSITION;
+        assembly {
+            s.slot := position
+        }
+    }
+}
+
+library PaymentVaultLib {
+    using SafeERC20 for IERC20;
+
+    event ReservationCreated(
+        uint64 indexed quorumId, address indexed account, PaymentVaultTypes.Reservation reservation
+    );
+
+    function s() internal pure returns (PaymentVaultStorage.Layout storage) {
+        return PaymentVaultStorage.layout();
+    }
+
+    function addReservation(
+        uint64 quorumId,
+        address account,
+        PaymentVaultTypes.Reservation memory reservation,
+        uint64 schedulePeriod
+    ) internal {
+        checkReservation(quorumId, account, reservation, schedulePeriod);
+
+        require(
+            reservation.startTimestamp >= s().quorum[quorumId].user[account].reservation.endTimestamp,
+            "Invalid start timestamp"
+        );
+        require(reservation.startTimestamp >= block.timestamp, "Invalid start timestamp");
+
+        populateReservedSymbols(
+            quorumId, reservation.startTimestamp, reservation.endTimestamp, reservation.symbolsPerSecond, schedulePeriod
+        );
+        s().quorum[quorumId].user[account].reservation = reservation;
+    }
+
+    function increaseReservation(
+        uint64 quorumId,
+        address account,
+        PaymentVaultTypes.Reservation memory reservation,
+        uint64 schedulePeriod
+    ) internal {
+        PaymentVaultTypes.Reservation storage currentReservation = s().quorum[quorumId].user[account].reservation;
+        checkReservation(quorumId, account, reservation, schedulePeriod);
+
+        require(reservation.startTimestamp == currentReservation.startTimestamp, "Invalid start timestamp");
+        require(reservation.endTimestamp > currentReservation.endTimestamp, "Invalid end timestamp");
```

Comment:
maybe use >= so increaseReservation don't necessarily mean to increase time range? Could we combine the check so that at least 1 of the time range or the symbols were increased? Like a separate check on `reservation.endTimestamp > currentReservation.endTimestamp || reservation.symbolsPerSecond > currentReservation.symbolsPerSecond`